File: Archive/psthBins.m
function  [r,rSmo,t] = psthBins(times, binsize, fs, ntrials, triallen)
% PSTHBINS Computes the peri-stimulus time histogram from spike times.
% 
% H = PSTH(TIMES, BINSIZE, FS,NTRIALS,TRIALLEN)
% TIMES - spike times (samples)
% BINSIZE - binwidth (ms)
% FS - sampling rate (hz)
% NTRIALS - number of trials
% TRIALLEN - length of a trial (samples)
%
% R - binned firing rate. 
% rSmo - smoothed firing rate (look inside function for smoothing window
% T - peri stimulus time histogram edges



% Author: Rajiv Narayan
% askrajiv@gmail.com
% Boston University, Boston, MA
%
% edited: [EHS::20211108]

%Compute PSTH        
lastBin = binsize * ceil((triallen)*(1000/(fs*binsize)));
edges = 0 : binsize : lastBin;	
x = (mod(times,triallen))*(1000/fs);
r = (histc(x,edges)*1000) / (ntrials*binsize);
t = edges;

% and smoothing
rSmo = smoothdata(r,'movmean',5);

File: Archive/psthBins.py
def psthBins(times, binsize, fs, n_trials, trial_len):
    '''
    psthBins computes the peri-stimulus time histogram from spike times (code adapted from Rajiv Narayan's psthBins.m).
    
    Inputs:
        - times: spike times (samples)
        - binsize: bin size (ms)
        - fs: sampling rate (Hz)
        - n_trials: number of trials
        - trial_len: length of each trial (samples)
    
    Outputs:
        - binned_FR: binned firing rate
        - smooth_FR: smoothed firing rate
        - T: peri-stimulus time histogram edges
    '''
    
    # Import libraries
    import numpy as np
    
    # Compute PSTH
    last_bin = binsize * np.ceil(trial_len / (1000 / (fs*binsize)))
    edges = np.arange(0, last_bin, binsize)
    x = np.mod(times, trial_len) * (1000 / fs)
    binned_FR = (np.histogram(x, edges) * 1000) / (n_trials * binsize)

    # Smoothing
    # smooth_FR = 
    
    return binned_FR, edges